x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/tree/Node.java

```diff
@@ -71,17 +72,25 @@ public void forEachDown(Consumer<? super T> action) {
 
     @SuppressWarnings("unchecked")
     public <E extends T> void forEachDown(Class<E> typeToken, Consumer<? super E> action) {
-        forEachDown(t -> {
-            if (typeToken.isInstance(t)) {
-                action.accept((E) t);
-            }
-        });
+        try {
```

Comment:
I can imagine that different rules will build different stacks, but also wondering if these functions aren't eventually called from smth like EsqlSession#parse() or similar?
I.e. wondering if we can get away with catching the error somewhere lower in the stack.

Reply:
They are called in multiple places, and some of them imply some async execution, so centralizing this check is not trivial; I tried to push it at a lower level actually (say at session level), but I ended up spreading this even more.